flipperdevices/flipperzero-firmware
# Bluetooth Remote Additions

# What's new

- Updated the Bluetooth Report Map to be a composite HID map
- Report now includes, keyboard, mouse and consumer controls
- Fixed the Media Player in the bluetooth remote (Volume Up will take camera captures on ios and android)
- Updated Bluetooth remote to add a mouse and keyboard remote
- Added a new bluetooth icon to save on space
- Long pressing back will allow a user to exit or change the remote
- Update usb hid keys to pull from libusb_stm32 library & update bad usb keys

# Notes

Due to the changes, old connections need to be forgotten and the device paired again. Changing the mac address for the profile does work, but to avoid confusion forgetting the old device is probably better.
file: furi_hal_bt.c
```
if(profile == FuriHalBtProfileHidKeyboard) {
	// Change MAC address for HID profile
	config->mac_address[2]+=2;
	...
```


# Verification  

- Launch the Bluetooth remote app and pair to a desktop or mobile device (android, for set full features)
- Ensure Long pressing the back button will open the submenu from any remote
	- Ensure clicking stay on the submenu will stay on the current remote
	- Ensure clicking ok on the submenu will open the remote selection menu
	- Ensure clicking back will exit the app
- Ensure navigation works with the keynote remote
	- Ensure ok will click space
	- Ensure back will click backspace & on mobile will backspace and click the back key
- Switch to the media player remote
	- Ensure up & down keys control volume
	- Ensure left and right keys control media tracking
	- Ensure ok button press will pause and play
- Switch to mouse remote
	- Ensure arrow keys can control the cursor
	- Ensure ok button can left click
	- Ensure long ok button press will turn on selection mode (click again to end)
	- Ensure back button will right click
- Switch to keyboard remote
	- Ensure arrow keys change the selected key
	- Ensure clicking the ok button on selected key will input that key
	- Ensure long ok button press will repeat the keys
	- Ensure clicking a modifier key, ctrl, alt, cmd (Win/Cmd), will both click the key & toggle the key
		- Ensure entering another character will use the toggled modifier(s)
	- Ensure back button will delete the last character
	- Ensure being disconnected will display the title and status, while connecting will show the keyboard
- Ensure bad usb scripts still work

# Checklist (For Reviewer)
